### ISSUE TITLE
Support JSON values that aren’t mapped

### ISSUE THREAD

**Issue by gkellogg** [w3c/json-ld-syntax#4]
* Consider using `”@type”: “@json” to describe native values in the compact form.
* Native values should include all JSON types: strings, booleans, numbers, and _null_ as well as objects and arrays.
* Expanded form can record these as values of `@value`.
  * Does interfere with some use of `[]` and `{}` in framing

Original issue is [Support JSON values that aren't mapped #333](https://github.com/json-ld/json-ld.org/issues/333)

**Comment by azaroth42** [comment]
:+1: This is a better (IMO) solution for native JSON values such as GeoJSON than requiring every community to map all of their constructs into -LD. 

To quote (with slight edits) the example from the original issue:

```
{
  "@context": {
    "@vocab": "http://example/",
    "@base": "http://example/",
    "json-value": {"@type": "@json"}
  },
  "@id": "foo",
  "json-value": {"native": "json"}
}
```

This seems very sensible, and fits with our charter. We can later make `@json` an alias for whatever literal type a future RDF WG might assign for JSON.

**Comment by akuckartz** [comment]
I would prefer a more LD friendly solution for GeoJSON. #7 ?

**Comment by azaroth42** [comment]
@akuckartz I didn't mean to imply that GeoJSON-LD was a bad thing to do, just that if the requirement is "support native JSON data structures in the JSON-LD context", then GeoJSON could be managed that way without then layering on GeoJSON-LD. GeoJSON-LD is great ... but if you don't need to interact with the -LD part of it, just record the JSON structure, there's overhead that could be minimized.  

There's a separate issue for the list of lists feature beyond #7 that was already accepted to be part of 1.1.  #7 would additionally let the semantics of the list of lists be expressed.

**Comment by gkellogg** [comment]
The key is the expanded form; my thought was that the previous example might expand to something like the following:

```json
[{
  "@id": "http://example/foo",
  "http://example/json-value": [{
    "@value": {"native": "json"},
    "@type": "@json"
  }]
}]
```

Regarding #7, this is not in conflict with a potentially more semweb-y mapping for GeoJSON, but there are other reasons why you might want to preserve raw JSON within JSON-LD.

When turned into RDF, we would need a datatype to describe the value, so that you would get something like the following:

```turtle
@base <http://example/foo> .
@prefix jsonld <https://www.w3.org/ns/json-ld#> .

<foo> <json-value> '{"native": "json"}'^^jsonld:json .
```

Where the JSON is normalized to use minimal whitespace.

**Comment by iherman** [comment]
I think defining a jsonld:json datatype woukd make a lot if sense at this day and age... and would offer a clean solution.

**Comment by davidlehn** [comment]
Should perhaps be jsonld:JSON to better align with https://www.w3.org/TR/rdf11-concepts/#section-htm

**Comment by azaroth42** [comment]
[WG resolved](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2018/2018-08-31-json-ld#resolution4) to add a @JSON keyword, mapped to jsonld:JSON to identify the JSON data type.

**Comment by BigBlueHat** [comment]
I'm concerned this opens a Pandora's box...or maybe several. Sadly, I wasn't here for the call and had overlooked this issue earlier, so I fear I'm only just now raising these concerns...

We're (rather passively) introducing a namespace specific to JSON-LD: `https://www.w3.org/ns/json-ld#`

We're inviting developers to avoid/ignore the graph model JSON-LD encodes:
```json
{
  "@context": {
    "data": {"@type": "@json"}
  },
  "data": {"everything": "imaginable"}
}
```

I fear providing this as a "solution for native JSON values such as GeoJSON" sends the wrong message...and it begins to invalidate the reason to have JSON-LD at all (see the example above).

Are we also planning to do this for YAML? Because the use cases would be identical...

**Comment by cwebber** [comment]
Having implemented the RDF canonicalization spec with a minor headache, this sounds like a full on migraine.

yo dawg I heard you like canonicalization so I put a tree data serialization canonicalization algorithm in your graph data serialization canonicalization algorithm so you can normalize while you normalize

**Comment by ajs6f** [comment]
@BigBlueHat I appreciate (and to some extent share) that concern, but I wonder if there's a historical analogy: I've not seen the kind of problem you are describing using XML literals within RDF/XML. That may not be a valid analogy, but it's a bit suggestive...

**Comment by azaroth42** [comment]
Re YAML, I don't think we would do that, because (a) no one has asked for it and (b) YAML is a non-normative deliverable of how the patterns of JSON-LD could be used in YAML to accomplish the same ends. The charter says:  "JSON-LD 1.1 examples specified in YAML" not a normative YAML-LD Rec.

We would be introducing a namespace, yes. We could also (as discussed on the call) add the data type to the RDF namespace, but we at least would need to document it. The consensus was that the creation of a new namespace was less work than putting it into an existing one, and a future RDF WG could take it over down the line.

I agree with @ajs6f about the use of [XML literals in RDF/XML](https://www.w3.org/TR/rdf-syntax-grammar/#section-Syntax-XML-literals). Yes, you can create pointless RDF that simply wraps a single literal in XML or JSON ... but why would you bother to do that? It seems like an enormous waste of your time other than to meet some badly worded RFP.

**Comment by gkellogg** [comment]
As @ajs6f points out, other RDF syntaxes that leverage languages have a similar mechanism for including raw XML or HTML, this is really no different.

For RDF canonicalization, such values would be treated just as other datatyped literals. Part of the RDF serialization aspects should include whitespace normalization, which is fairly standard in JSON, so I don't really appreciate why things such as RDF Dataset Normalization and signatures would be at any disadvantage.

**Comment by gkellogg** [comment]
@BigBlueHat worries about introducing a new namespace:

> We're (rather passively) introducing a namespace specific to JSON-LD: `https://www.w3.org/ns/json-ld#`.

In fact, this namespace already exists for URIs such as `http://www.w3.org/ns/json-ld#expanded` used in HTTP headers.

However, we don't need to use this namespace, and @iherman suggested that we could probably use the RDF namespace `http://www.w3.org/1999/02/22-rdf-syntax-ns#` and use `http://www.w3.org/1999/02/22-rdf-syntax-ns#JSON` as the datatype, making it first-class with `XMLLiteral` and `HTML` datatypes. Updating the RDF namespace document is something we can do, apparently.

I agree that this no longer serves for GeoJSON, and we should consider some other example, but such examples doubtless exist, which is why this is a compelling feature.

**Comment by Fak3** [comment]
> There's a separate issue for the list of lists feature beyond #7 that was already accepted to be part of 1.1.

@azaroth42 is there a github issue for the list of lists support? If not, may I create one?

**Comment by gkellogg** [comment]
@Fak3 The lists of lists issue is #36, and it was closed as support was added for recursive lists.

**Comment by cwebber** [comment]
> For RDF canonicalization, such values would be treated just as other datatyped literals. Part of the RDF serialization aspects should include whitespace normalization, which is fairly standard in JSON, so I don't really appreciate why things such as RDF Dataset Normalization and signatures would be at any disadvantage.

It isn't that simple.  Whitespace is not the only issue.  We will probably have to support something like [this json canonicalization spec](https://tools.ietf.org/html/draft-rundgren-json-canonicalization-scheme-01) or something.  That's a lot of extra work.

There's also a huge risk that people will open this loophole much, much wider than is anticipated, marking giant swaths of content as json-only.  Yeah, I guess that's true for XML too, but to be honest no sane person could operate on XML-RDF as if it were real XML and have things survive... it was an RDF serialization format and little more.  Here people are actually working with json-ld as if it were normal json and getting reasonable RDF interop.  There are pain points occasionally, and we should try to remedy those, but I think this is opening an escape hatch that a good number of people will jump straight through.

Careful about rubbing this lamp... I think fulfilling this wish will have more side effects than anticipated and may undo a lot of the goals of json-ld.  -1 from me.

**Comment by dlongley** [comment]
I share the same concerns as @cwebber and @BigBlueHat.

**Comment by azaroth42** [comment]
Re canonicalization (or even just whitespace normalization) ... can someone describe the issue and the risk here?  If one implementation serializes to a string `"{\"foo\": 1}"` and another serializes to a string `"{ \"foo\" : 1 }"` ... what's the problem? They're not identity providing such that they need to be compared, they're just values.

**Comment by cwebber** [comment]
@azaroth42 Those would end up being two different signatures with linked data signatures.  Without canonicalizing the json exactly the same way every time, LDS will break.

**Comment by azaroth42** [comment]
We can defer, but I would like to note that canonicalization and LDS are explicitly out of scope of the WG, per the charter:  https://www.w3.org/2018/03/jsonld-wg-charter.html

**Comment by azaroth42** [comment]
PROPOSAL:  Defer work on JSON literals until specific use cases have been described.

**Comment by BigBlueHat** [comment]
> We can defer, but I would like to note that canonicalization and LDS are explicitly out of scope of the WG, per the charter: https://www.w3.org/2018/03/jsonld-wg-charter.html

They are out of scope, but we should still be careful not to break them.

**Comment by iherman** [comment]
Thanks @cwebber for the [json canonicalization spec](https://tools.ietf.org/html/draft-rundgren-json-canonicalization-scheme-01) link. I did not know about that. 

However, that also means that the signature issue with a json literal becomes solvable if and when an RDF canonicalization is formally defined as a standard; that standard should say that literals must be canonicalized and, actually, by defining the json datatype we _reinforce_ interoperability of signatures because there is then a clear follow-your-nose approach of what this means in the standard (provided the aforementioned IETF document becomes final). Indeed, _if_ there are use cases that would make all kinds of ad-hoc inclusion of JSON data in RDF graphs but without a clear spec, a canonicalization spec may not be able to function properly.

Ie, while I agree that we need more use cases per our process, I think a future RDF canonicalization spec can take care of this problem for us, and this WG does not really have to deal with this.

(B.t.w., does the XML Canonicalization spec apply to HTML content? Because if not, there is already a problem with the HTML RDF datatype...)

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2018/2018-09-07-json-ld).

- `RESOLVED:  Defer work on JSON literals until specific use cases have been described`
<details><summary><i class="differentiate">View the transcript</i></summary>
Benjamin Young: <a href="https://github.com/w3c/json-ld-syntax/issues/4">https://github.com/w3c/json-ld-syntax/issues/4</a><br/>
Benjamin Young: <a href="https://github.com/w3c/json-ld-syntax/issues/4#issuecomment-418857569">https://github.com/w3c/json-ld-syntax/issues/4#issuecomment-418857569</a><br/>
Benjamin Young: first issue is number 4. there had been an agreement in past meeting to close but it was brought back up. Further discussion on GitHub and there is now a proposal from Rob to defer. Want to open it back up for further discussion<br/>
Proposed resolution: Defer work on JSON literals until specific use cases have been described (Benjamin Young)<br/>
Gregg Kellogg: chris weber brought up issue of canonical order of the literal. there is a draft out for JSON canonicalization but it is a burden until algorithms there to do it. This issue is related to the RDF JSON canonicalization. we have this issue with HTML and XML literals.<br/>
… we are laking a compelling use case for this<br/>
… part was to support geo json but we added list of list to support that<br/>
Adam Soroka: to clarify, geo json has no use case for this current issue?<br/>
Rob Sanderson: Related to #7 – <a href="https://github.com/w3c/json-ld-syntax/issues/7">https://github.com/w3c/json-ld-syntax/issues/7</a><br/>
Gregg Kellogg: list of list allows us to represent geo json but some of the proponents of geo json want to see us go further<br/>
Adam Soroka: happy with the proposal and to see conversation to move forward. We owe it to geo json to see that their issues are represented in our work<br/>
Rob Sanderson: while I supported the issue, given the lack of use cases, I support the deferral.<br/>
David Newbury: I glad it is a deferral and not a close. but when you need to semanticalize and de-semanticalize json it can be a pain.<br/>
Ivan Herman: no strong opinion about literals in json but do want to make sure we do not close it for the wrong reason. understand the canonicalization issue but it is not our responsibility. should we have a more broad RDF canonicalization group.<br/>
Rob Sanderson: +1 to ivan.<br/>
Adam Soroka: +1<br/>
David I. Lehn: i’ll just note, that similar to empty terms, raw json would be very helpful for obfuscated json-ld<br/>
Benjamin Young: csvw:JSON<br/>
Benjamin Young: did more digging and there is another community that minted a json encoding type. the csv working group has this. The json is not native json but rather just a string but has way to say, this is a json string.<br/>
… need to make sure we have a compelling use case<br/>
Rob Sanderson: +1<br/>
Benjamin Young: does anyone want changes?<br/>
Proposed resolution: Defer work on JSON literals until specific use cases have been described (Benjamin Young)<br/>
David Newbury: +1<br/>
Jeff Mixter: +1<br/>
Ivan Herman: +1<br/>
Adam Soroka: +1<br/>
Benjamin Young: +1<br/>
Harold Solbrig: +1<br/>
Gregg Kellogg: +1<br/>
Simon Steyskal: +1<br/>
Resolution #2: Defer work on JSON literals until specific use cases have been described<br/>
David I. Lehn: +1<br/>
</details>

**Comment by azaroth42** [comment]
Use Case:  In the CIDOC-CRM vocabulary, there is a property that takes a Literal for the description of a Place.  This (thus) can take a lat,long pair as a string, or an XmlLiteral ... but not GeoJSON as a literal, as there is no JSON datatype. With a JSON data type, and support in JSON-LD for `@type:@json` in a term definition, we could thus embed the GeoJSON directly in the document, rather than unnecessarily serializing to a string.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-02-07-json-ld#section4).

- `RESOLVED:  Add JSON data type to RDF, with support in JSON-LD processors for managing parsed JSON in the internal form. We will seek feedback via blogpost, and in Berlin. {: #resolution9 .resolution}`
<details><summary><i class="differentiate">View the transcript</i></summary>
JSON Literal<br/>
Harold Solbrig: <a href="https://github.com/w3c/json-ld-syntax/issues/4">https://github.com/w3c/json-ld-syntax/issues/4</a><br/>
Ivan Herman: to do this, we will have to define an RDF data type for JSON<br/>
Ivan Herman: look at RDF concepts document - 3 or 4 lines<br/>
Rob Sanderson: <a href="https://www.w3.org/TR/rdf11-concepts/#section-html">https://www.w3.org/TR/rdf11-concepts/#section-html</a><br/>
Gregg Kellogg: what we need to do that those didn’t is to deal with white space…<br/>
Ivan Herman: the only thing we have to do is to say there is a string, declared to be JSON<br/>
Ivan Herman: it makes cannonicalization difficult so we’d have to find or referenc it<br/>
… the fact that there are html data types that makes cannonicalization difficult or impossible<br/>
… so if you want to do cannonicalization, html is the big kahuna, and json is secondary<br/>
… you may want to answer question about json equality, but is that our responsibility.<br/>
Gregg Kellogg: when you parse html or xml, value elements preserve white space. As a result, indentation variations give different literals.<br/>
… in json, parsing is done after parser is completed and parsers don’t preserve white space. Serializers can be told no unnecessary white space<br/>
… which avoids html and xml problem and allows us to be relatively immune. Ordering may be different but …<br/>
… parsers aren’t required to be order preserving, so literals from 2 parsers might not compare. We could state that they appear<br/>
… equivalent as objects in testing infrastructure.<br/>
… so we just need to state that values aren’t stored without extra white space.<br/>
Ivan Herman: not sure that specification should require that when not a question of testing.<br/>
Ivan Herman: question is whether it is up to us to define when two pieces of JSON are equal<br/>
Ivan Herman: we can be very pragmatic and say that someone else has to take care of this.<br/>
Gregg Kellogg: I think we need to say something about this. We need to say something about how that serialization is performed.<br/>
… we need to say how you create that string from the objects.<br/>
Adam Soroka: consider the case of GeoJSON …<br/>
David I. Lehn: is there another solution than complete serialization?<br/>
Rob Sanderson: Use case p168 property has range rdf literal with all sorts of possible values, but you can’t add a json literal because there is no id.<br/>
… if we want to make it easy for authors to do the right thing, then json as (readable) json is what is needed.<br/>
Harold Solbrig: <a href="http://build.fhir.org/medicationexample0301.json.html">http://build.fhir.org/medicationexample0301.json.html</a><br/>
Harold Solbrig: <a href="http://build.fhir.org/medicationexample0301.ttl.html">http://build.fhir.org/medicationexample0301.ttl.html</a><br/>
Rob Sanderson: spec above is completely silent about html canonicalization<br/>
… we aren’t adding a new problem, but just another type that doesn’t define canonicalization<br/>
Rob Sanderson: can we punt on it and say will be canonicalized when there is a canonicalizatin spec for json?<br/>
Ivan Herman: in publishing we put out jsonld not for any reason except it has been accepted by schema.org<br/>
… any author who cares about being found by schema.org will put it in jsonld vs. json literal<br/>
David Newbury: this will allow people to progressively add semantics vs. having to do everything up front<br/>
Benjamin Young: doing this without canonicalization is painfully naive and dangerous. If isn’t consistent between python and js…<br/>
… we’re just asking for a world of pain.<br/>
Rob Sanderson: question is the extent to which … (if we don’t do complete canonicalization…) … are we doing a better or worse job than nothing?<br/>
Gregg Kellogg: coming around to saying it is undefined until spec exists.<br/>
Gregg Kellogg: signature specs will need to call out some canonicalization spec or define one that should be used by jsonld processors …<br/>
… suggest that canonicalization of jsonld literals is not supported, so can’t sign graphs that contain it…<br/>
David I. Lehn: is there a way to know which ones aren’t supported?<br/>
Ivan Herman: no because anyone can define a datatype in an RDF graph.<br/>
David Newbury: unless codepaths do unicode canonicalization, we’ve still got issues…<br/>
Ivan Herman: are there good reasons to do this in jsonld?<br/>
Ivan Herman: do we want to have json literals or not?<br/>
David I. Lehn: I’m worried about people misusing it but…<br/>
Gregg Kellogg: if there WAS a standard for it, we’d say it must be in that form …<br/>
Rob Sanderson: there are many terrible things people can do today. I’m highly reluctant to use “it could be misused” as a good reason to not do something.<br/>
David Newbury: when we talk about misuse, do we mean security or misuse of RDF in the world (decreasing the amount of semantics in the universe)<br/>
Ivan Herman: From a W3C perspective… what we could do is put this into the document w/ a note that says it could be done but we want to have…<br/>
… feedback from the community that we are not sure that it is a good idea and seek input.<br/>
Gregg Kellogg: we did something like this in spec w/ bnode identifiers and document base url …<br/>
Jeff Mixter: reason that WikiData chose JSON rather than JSONLD … they may have opinions about injecting into … document.<br/>
Proposed resolution: Add JSON data type to RDF, with support in JSON-LD processors for managing parsed JSON in the internal form. We will seek feedback via blogpost, and in Berlin. (Rob Sanderson)<br/>
Ivan Herman: +1<br/>
Gregg Kellogg: +1<br/>
Rob Sanderson: +1<br/>
David Newbury: +1<br/>
Jeff Mixter: +1<br/>
David I. Lehn: +0.9<br/>
Harold Solbrig: +1<br/>
Resolution #8: Add JSON data type to RDF, with support in JSON-LD processors for managing parsed JSON in the internal form. We will seek feedback via blogpost, and in Berlin. {: #resolution9 .resolution}<br/>
</details>

**Comment by gkellogg** [comment]
For basic interoperability, and to allow implementations to pass tests, I believe we will  need to tackle canonicalization. The current version of the draft is https://tools.ietf.org/html/draft-rundgren-json-canonicalization-scheme-05, not sure when it's likely to become an RFC, so don't think we can normatively cite it (although, it looks like it will become an RFC and the core bits aren't too likely to change). We might just need to re-describe the normalization points and informatively reference the draft:

* No whitespace between JSON tokens.
* Within strings, code points between within the traditional ASCII control
      character range (`U+0000` through `U+001F`), it **MUST** be serialized
      using lowercase hexadecimal Unicode notation (`\uhhhh`) unless it is
      in the set of predefined JSON control characters `U+0008`, `U+0009`,
      `U+000A`, `U+000C` or `U+000D` which **MUST** be serialized as `\b`, `\t`, `\n`,
      `\f` and `\r` respectively. If the Unicode value is outside of the ASCII control character
      range, it **MUST** be serialized "as is" unless it is equivalent to
      `U+005C` (`\`) or `U+0022` (`"`) which **MUST** be serialized as `\\` and `\"`
      respectively.
* JSON Number data **MUST* be serialized according to Section 7.1.12.1 of
   [[ES6](https://www.ecma-international.org/ecma-262/6.0/index.html)] including the "Note 2" enhancement.
* JSON Object properties MUST be sorted in a recursive manner which
      means that possible JSON child Objects MUST have their properties
      sorted as well.

There are [implementations available](https://github.com/cyberphone/json-canonicalization) for several languages (not Ruby, which I'll work on and contribute).

**Comment by iherman** [comment]
On https://github.com/w3c/json-ld-syntax/issues/4#issuecomment-474030501: I am not sure, @gkellogg.

- As referred to in the comments, the fact of introducing an RDF datatype does not _require_ to define a canonical format. Per [RDF1.1 spec](https://www.w3.org/TR/rdf11-concepts/#section-Datatypes):

    > A datatype consists of a lexical space, a value space and a lexical-to-value mapping, and is denoted by one or more IRIs.

    More importantly, there is an example for a datatype (i.e., rdf:HTML) that does _not_ have a canonical version for now. I understand this creates problems  and, I presume, this means that an RDF graph containing an HTML (or JSON) literal _cannot_ be properly signed, but the question is whether this creates for the use cases that we have for JSON literals. Similarly, I do not see why you claim

    > For basic interoperability, [...], I believe we will need to tackle canonicalization.
- We may indeed need this for pass tests. But that is a fundamentally different problem, which does not require us to include a _recommendation_ for a canonicalization. We can just refer to the [IETF document](https://tools.ietf.org/html/draft-rundgren-json-canonicalization-scheme-05).
- _If_ we really need something normative in our document, we should try to find out what exactly is the status of that document. _If_ it is a stable document (and we can have some sort of a clear statement from somebody of authority at IETF), then we may be able to convince the director to allow us a normative reference. That would be much better than us defining a normative canonicalization that may become in conflict with others.

**Comment by pchampin** [comment]
@iherman JSON literals in JSON-LD might be more tricky than other literals (including HTML) in RDF. The difference is that, in other RDF serializations, literals have a fixed lexical representation. Unless some form of inference on literals is performed, this lexical representation is kept as is when processing the graph.

In JSON-LD, the JSON literal we get (*e.g.* in the expansion algorithms) is *not* a lexical representation, it is the *value* itself (a JSON object). When converting to triples (and possibly to the expanded form?), a lexical representation has to be built, and there is no unique way to do that. I think that's what @gkellogg means by "basic interoperability problems".

**Comment by iherman** [comment]
@pchampin 

> In JSON-LD, the JSON literal we get (e.g. in the expansion algorithms) is not a lexical representation, it is the value itself (a JSON object). 

I am not sure that is a given. Going back to the RDF datatype definition: what would exactly be the [value space](https://www.w3.org/TR/rdf11-concepts/#dfn-value-space) of a JSON Literal? Maybe even more specifically: what is the definition of equality in such a space? The former _is_ something we must define if we define an RDF Datatype. Referring to the canonical representation might be indeed used for the definition of the value space, but that is not necessarily the only way to do that. We could (just from the top of my head) say that a JSON object (or an array thereof) is parsed into a JavaScript object using the JSON parsing rules, and the result is the 'value space'. Which means that two JSON objects are equal if their JavaScript objects are equal per the rules defined by JavaScript.  I am not saying that the JavaScript approach is the right one, but it has the value of relying on a standard. 

*If* we have such a clear value space (with equality on it) then

> When converting to triples (and possibly to the expanded form?), a lexical representation has to be built, and there is no unique way to do that

is not a problem anymore.

To be clear, I do not have a clear answer. But I am worried about copying an IETF document and making it part of a recommendation. That is what I would prefer to avoid...

**Comment by gkellogg** [comment]
@cyberphone, can you comment on the status of 
https://tools.ietf.org/html/draft-rundgren-json-canonicalization-scheme-05? If we are to support JSON literals, it would be best to canonicalize them. When is this expected to become an RFC? How stable is the document? Are there other specs which are normatively referencing the spec?

@iherman part of testing requires an RDF transform and using dataset isomorphism. At that point, the precise lexical representation of JSON literals becomes important. Certainly, this could be left out of the spec, and used in test-suite instructions, but for many reasons, setting on a canonical form for JSON literals is going to be important, if we can overcome the normative citation issues.

**Comment by gkellogg** [comment]
My Ruby version for JSON canonicalization: https://github.com/dryruby/json-canonicalization.

**Comment by cyberphone** [comment]
@gkellogg It is great to see a sixth incarnation of the proposal!

Regarding progress the technical issues have (AFAICT...) been properly identified; the problem is rather that a bunch of people still consider canonicalization as pure stupidity.  OTOH, it seems that none of the current Open Banking APIs has bought into the Base64Url-concept either.

FWIW, I will do a short presentation 
https://cyberphone.github.io/ietf-signed-http-requests/hotrfc-shreq.pdf
at IETF-104 in Prague which shows how you can apply JCS on a mainstream application.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-03-22-json-ld#section1-3).

- `RESOLVED:  Move forwards with a JSON native data type, with a warning that it cannot be canonicalized`
<details><summary><i class="differentiate">View the transcript</i></summary>
JSON datatype<br/>
Rob Sanderson: link: <a href="https://github.com/w3c/json-ld-syntax/issues/4">https://github.com/w3c/json-ld-syntax/issues/4</a><br/>
Rob Sanderson: PR: <a href="https://github.com/w3c/json-ld-api/pull/72">https://github.com/w3c/json-ld-api/pull/72</a><br/>
Rob Sanderson: we also have discussed the JSON datatype on github<br/>
… Gregg, you’ve been the most involved (as always)<br/>
… could you summarize?<br/>
Gregg Kellogg: the issue comes down to representation<br/>
… if you are going to describe both the lexical and value space<br/>
… somewhat like HTML<br/>
… the lexical space cannon be guaranteed<br/>
… the JSON literal quality is lost when its turned into a native representation<br/>
… you loose the original key ordering, key escaping, and lexical numerical representations<br/>
… so it seems we will need to canonicalize<br/>
… which has been referenced in the issue<br/>
… it’s sadly not as close to done as I’d hoped<br/>
… and we can’t count on it being final in time<br/>
… so, do we care if two implementations use the same canonicalization<br/>
… so we have done some things about do we use Integer or Doubles for numbers<br/>
… so when you’d turn the JSON literal into RDF (in the toRDF space), we do need to say something about that at least<br/>
… and the elimination of whitespace<br/>
… and the ordering of keys<br/>
… I think that can be done<br/>
… there’s a lot of detail in that, but we should be able to reference ECMAScript for this<br/>
… or we could do it ourselves<br/>
Rob Sanderson: last time we talked about the canonicalization issue<br/>
… we also talked about HTML being not easily canonicalizable<br/>
Gregg Kellogg: HTML is a little different<br/>
… they will preserve order, and whitespace<br/>
… so you do have the opportunity return to that result<br/>
Ivan Herman: well, attribute order and things are not covered<br/>
… this would be a problem if you were to attempt to sign an HTML document<br/>
Gregg Kellogg: if we weren’t in an era when signatures weren’t as important as they are now, then maybe we wouldn’t need to care about this so much<br/>
Rob Sanderson: so, is there a JSON-LD document that could include a JSON “native” data type that also needs to be signed<br/>
… so if the only use case is to import GeoJSON<br/>
… do we need to worry<br/>
Ivan Herman: I have spent time on this issue with others<br/>
… aside from the canonicalization problem<br/>
… if we do make a native JSON type, we will have to put it into some namespace–rdf: or jsonld:<br/>
Rob Sanderson: +1 to RDF namespace<br/>
Ivan Herman: if we do that, we’ll have to write the SWIG mailing list, to announce the new datatype, etc.<br/>
… we can do this as part of our document<br/>
… the other problem is<br/>
… I did put a reference in the issue for the rules we have to follow when we point to something normatively<br/>
… my first reading is that unfortunately, this JSON canonicalization specification cannot be referred to normatively<br/>
… the second problem is bringing our own canonicalization into our document<br/>
… if we do that, I can safely say the Director would say no to that<br/>
… so, we can’t just take an IETF spec and put it into a W3C spec<br/>
… all of these are admin problems<br/>
… But I am still not convinced that we need the canonicalization as a normative part of our spec<br/>
… we could say that someone else may do this and reference forthcoming work<br/>
… but when the issue is that we have a JSON portion we want to store in RDF<br/>
… we can state that the only expectation is that [the same processor will produce the same output]<br/>
… none of the arguments that I heard is that canonicalization needs to be normative<br/>
Pierre-Antoine Champin: <a href="http://tinyurl.com/y2gmzxf8">http://tinyurl.com/y2gmzxf8</a><br/>
Pierre-Antoine Champin: I was wondering about this example<br/>
… there’s an Integer in the non-canonical form<br/>
… would that be canonicalized or not?<br/>
Gregg Kellogg: yes, that would be canonicalized<br/>
… I don’t know any processors that would properly serialize that with a leading zero<br/>
… if you’re going to the internal representation<br/>
… it is the number 42<br/>
… some might do 42.0<br/>
… or 42E+0<br/>
… that would be fine, but I don’t think most JSON serializers would do that<br/>
Pierre-Antoine Champin: for the moment, we know how to sign this thing<br/>
Dave Longley: I think this falls into the same category as HTML<br/>
… it’s a string in the JSON; it’s not native HTML<br/>
… or a native number in the example’s case<br/>
… if we’re storing stuff in a string, then store it as a string<br/>
… but people want a native JSON object in their JSON<br/>
Pierre-Antoine Champin: but if you remove the leading 0 you don’t get the same signature<br/>
… so I’m assuming that the signature is dealing with the order or absence of order in the object when signed<br/>
… so if the object was a native JSON object, then it would already benefit<br/>
… and regardless we already have this problem with other string-expressed literals<br/>
Rob Sanderson: if you instead make it value 42.0<br/>
… since no one really serializes as 042<br/>
… whatever you change here will change the signature<br/>
… even though it will canonicalize as something different<br/>
Dave Longley: I disagree<br/>
Rob Sanderson: what do you disagree with?<br/>
Ivan Herman: I think in these examples, the current JSON-LD specification doesn’t say anything about what you put in strings<br/>
… we don’t suggest any sort of mini-canonicalization for things like this<br/>
… having built-in canonicalization for the native JSON representation<br/>
… would be a departure from what we’ve done previously<br/>
Dave Longley: my response to all that is that we have very consistent rules about moving non-string data into strings<br/>
… so we do have those sorts of specifications<br/>
… from a native JSON value into a string<br/>
… this same thing would exist for native JSON objects<br/>
… for things that come in via a string, those will stay as whatever that string is<br/>
… so strings have no issue<br/>
… so if you take pchampin’s example, and change it to a real number: 42<br/>
Gregg Kellogg: 42, 42.0, 42.0E0, 4.2E+1 are all the same number<br/>
Dave Longley: and if you put that in the playground, check the nquads tab, you’ll find the same number<br/>
Ivan Herman: yep I acknowledge that<br/>
Rob Sanderson: maybe then it’s the playground which is at fault<br/>
… I put in several examples, and the signature changes for all of these different 42’s as an integer<br/>
Dave Longley: you’re looking at the RSA signature, so you’ll see it change constantly<br/>
… because that injects random data<br/>
… what you need to look at is the N-Quads or normalized tabs<br/>
… the data there stays the same<br/>
Gregg Kellogg: this is in the data round tripping section<br/>
Gregg Kellogg: so, imo, if we create a datatype for JSON<br/>
… before there is a canonicalization for it<br/>
… then we’re in danger of doing things too early<br/>
… ultimately we need to deal with a canonicalized JSON<br/>
Pierre-Antoine Champin: +1<br/>
Gregg Kellogg: so the best thing we can do right now is nothing<br/>
… and defer this until there is a canonicalized form<br/>
… otherwise whitespace, object ordering, etc are all variable<br/>
… and the literals really won’t be worth doing any lexical representation is important<br/>
… better not to do anything until a canonicalization spec exists<br/>
Ivan Herman: my take would be milder<br/>
… the GeoJSON example doesn’t care about canonicalization<br/>
Rob Sanderson: +1 to ivan<br/>
Ivan Herman: with the canonicalization things differed<br/>
… and state that this feature is not recommended<br/>
… so we differ it, and if/when the canonicalization becomes standard or whatever, then we at that point suggest that that spec gets used<br/>
Rob Sanderson: it would be better to have a JSON datatype and state that later we’ll do canonicalization<br/>
Dave Longley: let’s provide rules for how to produce the JSON string that match the draft – but that you can do something else and be very clear it’s preferred that everyone do the same thing<br/>
Rob Sanderson: so we should start with JSON datatypes, and just suggest that you can’t sign these<br/>
Jeff Mixter: +1 to ivan and azaroth<br/>
Gregg Kellogg: if we don’t do canonicalization now, we don’t seem to be prevented from doing it later<br/>
… if we end up as a living spec, then we could do it that way<br/>
… and we could also suggest that for testing purposes it is always canonicalized<br/>
Rob Sanderson: a warning or a note?<br/>
Proposed resolution: Move forwards with a JSON native data type, with a warning that it cannot be canonicalized (Rob Sanderson)<br/>
Rob Sanderson: I’d suggest a warning<br/>
Gregg Kellogg: +1<br/>
Jeff Mixter: +1<br/>
Ivan Herman: +1<br/>
Rob Sanderson: +1<br/>
Simon Steyskal: +1<br/>
Pierre-Antoine Champin: +1<br/>
Tim Cole: +1<br/>
Dave Longley: +0<br/>
Benjamin Young: +0 still have concerns about eager misuse<br/>
David I. Lehn: +0.5<br/>
Jeff Mixter: I echo bigbluehat concerns but I also have very valid reasons to add JSON to RDF data.<br/>
Dave Longley: +1 to everything Benjamin is saying … but that we should really also have JSON literals … but they should also all be converted to the same strings in processors :)<br/>
David Newbury: +1<br/>
Resolution #3: Move forwards with a JSON native data type, with a warning that it cannot be canonicalized<br/>
Dave Longley: JSON literals can be an escape hatch but ONLY an escape hatch.<br/>
</details>

**Comment by azaroth42** [comment]
Agree done, closing :)
